support/default_output_validator.cpp

```diff
@@ -2,199 +2,379 @@
 #include <sstream>
 #include <string>
 #include <vector>
-using namespace std;
 
+//============================================================================//
+// Constants                                                                  //
+//============================================================================//
+constexpr int AC = 42;
+constexpr int WA = 43;
+
+constexpr std::string_view CASE_SENSITIVE           = "case_sensitive";
+constexpr std::string_view SPACE_CHANGE_SENSITIVE   = "space_change_sensitive";
+constexpr std::string_view FLOAT_ABSOLUTE_TOLERANCE = "float_absolute_tolerance";
+constexpr std::string_view FLOAT_RELATIVE_TOLERANCE = "float_relative_tolerance";
+constexpr std::string_view FLOAT_TOLERANCE          = "float_tolerance";
+constexpr std::string_view TEXT_ELLIPSIS            = "[...]";
+constexpr std::string_view WHITESPACE               = " \f\n\r\t\v";
+
+//============================================================================//
+// parameters                                                                 //
+//============================================================================//
 bool case_sensitive         = false;
 bool space_change_sensitive = false;
+bool compare_floats         = false;
 
-double float_relative_tolerance = 0;
-double float_absolute_tolerance = 0;
-double float_tolerance          = 0;
+long double float_relative_tolerance = -1;
+long double float_absolute_tolerance = -1;
 
-string strip_newline(string s) {
-	if(not s.empty() and s.back() == '\n') s.pop_back();
-	return s;
-}
+//============================================================================//
+// Utility                                                                    //
+//============================================================================//
+namespace util {
+	constexpr bool is_space(char c) {
+		return WHITESPACE.find(c) != std::string_view::npos;
+	}
+
+	constexpr bool is_digits(std::string_view token) {
+		for (char c : token) {
+			if (c < '0' or c > '9') return false;
+		}
+		return true;
+	}
+
+	constexpr bool is_integer(std::string_view token) {
+		if (token.substr(0, 1) == "-") token.remove_prefix(1);  // ignore optional - sign
+		if (token.empty()) return false;                        // integers need at least one digit
+		if (token.size() > 1 and token[0] == '0') return false; // integers dont start with 0x
+		if (not is_digits(token)) return false;                 // integers are just digits
+		return true;
+	}
 
-string crop_output(string output) {
-	if(output.size() > 200) {
-		output.resize(200);
-		output += " ...";
+	constexpr bool is_decimal(std::string_view token) {
+		std::size_t dot = token.find('.');
+		if (not is_integer(token.substr(0, dot))) return false;                        // decimals before the dot are *non empty* integers
+		if (dot < token.size() and not is_digits(token.substr(dot + 1))) return false; // decimals only have digits after the dot
```

Comment:
This allows trailing dot (`0.`). Should we? I don't see a testcase to either AC or WA this, so that can be added.

Reply:
yes, trailing dot is typically considered ok.